packages/plugin-css/src/utils/generate-token-name.ts

```diff
@@ -0,0 +1,49 @@
+import type {ParsedToken} from '@cobalt-ui/core';
+
+function normalizeIdSegment(inputString: string): string {
+  const words = inputString.trim().split(' ');
+
+  /* Only camelCase id segments if they have a middle space.
+  This prevents a breaking change to names that have capitals but no spaces. */
+  if (words.length > 1) {
+    return words
+      .map((word, i) => {
+        if (i === 0) {
+          return word.toLowerCase();
+        } else {
+          return word[0]?.toUpperCase() + word.slice(1).toLowerCase();
+        }
+      })
+      .join('');
+  }
+
+  return words[0]!;
+}
+
+export const DASH_PREFIX_RE = /^-+/;
+const DASH_SUFFIX_RE = /-+$/;
+
+export function defaultNameGenerator(
+  /** dot separated path to token with possible type specific modifications (such as appended composite token property names)  */
+  variableId: string,
+  /** @deprecated implement prefixes directly in `generateName` instead */
+  prefix?: string,
+): string {
+  const normalizedPrefix = prefix ? `${prefix.replace(DASH_PREFIX_RE, '').replace(DASH_SUFFIX_RE, '')}-` : '';
+  return `${normalizedPrefix}${variableId.split('.').map(normalizeIdSegment).join('-')}`;
+}
+
+export type CustomNameGenerator = (variableId: string, token?: ParsedToken) => string;
+
+// TODO: remove prefix arg in next major version
+export function makeNameGenerator(customNameGenerator?: CustomNameGenerator, prefix?: string) {
+  return (
+    /** dot separated path to token with possible type specific modifications (such as appended composite token property names)  */
+    variableId: string,
+    token?: ParsedToken,
+  ): string => {
+    const name = customNameGenerator ? customNameGenerator(variableId, token) : defaultNameGenerator(variableId, prefix);
```

Comment:
I like your implementation, but I will make one tiny change here in a patch release: we call a user’s `customNameGenerator()` on every token, and they’re allowed to return `undefined` or `null` if they’d like to defer naming back to us.

This is common plugin behavior ([example: Rollup](https://rollupjs.org/plugin-development/#a-simple-example)) and is a super minor detail that doesn’t affect the overall code. Just a tiny logical change that rather than checking a user passed the function, instead check the **return value** of the function and fall back to `defaultNameGenerator()` if `customNameGenerator()` executed but returned a falsy value.

Reply:
The breakage was really impossible to test for; it had to do with mismatching versions of plugin-css and plugin-sass. And since we didn’t indicate this as a breaking change, we shouldn’t require both be upgraded (I didn’t realize it either). Not really a way to test for this.

So note I’m not changing any code you wrote or how this function works; just going to release a `patch` for plugin-css and plugin-sass that changes some internals so both can work correctly (which, ironically, downloading the patch would fix the problem anyway, but it at least signals to consumers that a problem was identified and fixed).

Again, no one’s fault, not a big deal, and just sort of a weird, unpredictable, hard-to-test-for scenario. But wanted to leave a note.